ingestion-beam/src/main/java/com/mozilla/telemetry/decoder/rally/DecryptRallyPayloads.java

```diff
@@ -0,0 +1,244 @@
+          if (destinationParent.isObject()) {
+            ((ObjectNode) destinationParent).set(
+                mapping.decrypted_field_path().last().getMatchingProperty(),
+                Json.readObjectNode(decrypted));
+          } else {
+            throw new DecryptRallyPayloadsException(
+                "missing parent object for destination field: " + mapping.decrypted_field_path());
+          }
+        }
+      }
+    }
+
+    @Override
+    public Iterable<PubsubMessage> apply(PubsubMessage message)
+        throws IOException, JoseException, KeyNotFoundException, IllegalArgumentException {
+      message = PubsubConstraints.ensureNonNull(message);
+      Map<String, String> attributes = new HashMap<>(message.getAttributeMap());
+      final String namespace = attributes.get(Attribute.DOCUMENT_NAMESPACE);
+
+      if (keyStore == null) {
+        // If configured resources aren't available, this throws
+        // UncheckedIOException; this is unretryable so we allow it to bubble up
+        // and kill the worker and eventually fail the pipeline.
+        keyStore = KeyStore.of(metadataLocation.get(), kmsEnabled.get());
+      }
+
+      if (pipelineMetadataStore == null) {
+        pipelineMetadataStore = PipelineMetadataStore.of(schemasLocation.get(),
+            BeamFileInputStream::open);
+      }
+
+      if (gleanSchema == null || validator == null) {
+        JSONSchemaStore schemaStore = JSONSchemaStore.of(schemasLocation.get(),
+            BeamFileInputStream::open);
+        gleanSchema = schemaStore.getSchema("glean/glean/glean.1.schema.json");
+        validator = new JsonValidator();
+      }
+
+      final PipelineMetadata meta = pipelineMetadataStore.getSchema(attributes);
+      if (meta.jwe_mappings().isEmpty()) {
+        // Error for lack of a better behavior
+        throw new DecryptRallyPayloadsException(
+            String.format("jwe_mappings missing from schema: %s %s", namespace,
+                attributes.get(Attribute.DOCUMENT_TYPE)));
+      }
+
+      ObjectNode json = Json.readObjectNode(message.getPayload());
+
+      // Decrypt in-place
+      try {
+        // The keying behavior is specific to Rally -- it may be possible to
+        // rely on the keyId in the header instead. This behavior should be
+        // consistent with how keys are actually allocated.
+        PrivateKey key = keyStore.getKeyOrThrow(namespace);
+        decryptAndReplace(key, json, meta.jwe_mappings());
+      } catch (JoseException | KeyNotFoundException e) {
+        throw e;
+      }
+
+      // Ensure our new payload is a glean document
+      validator.validate(gleanSchema, json);
+
+      // Ensure that our payload has a uuid field named rally.id, otherwise
+      // throw this docuemnt into the error stream.
+      final JsonNode rallyId = json.at("/metrics/uuid/rally.id");
+      if (rallyId.isMissingNode()) {
+        throw new DecryptRallyPayloadsException("missing field: #/metrics/uuid/rally.id");
```

Comment:
@Dexterp37 Is the naming of the uuid field here consistent with the one used in the rally-core addon (and any subsequent study down the line)?

Reply:
[Yes](https://github.com/mozilla-rally/rally-core-addon/pull/505/files?file-filters%5B%5D=.cjs&file-filters%5B%5D=.js&file-filters%5B%5D=.json&file-filters%5B%5D=.yaml#diff-320dfd9e9980a24eb3e544bc76582626b0f863ba93f8e042d45c7feb90af2a50R16-R17).